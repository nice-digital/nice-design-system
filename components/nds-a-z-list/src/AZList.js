import React from "react";
import PropTypes from "prop-types";
import "../scss/a-z-list.scss";

export const AZList = props => {
	const { alphabet: Alphabet, children, className, ...attrs } = props;

	return (
		<>
			<Alphabet className="a-z-list__alphabet" />
			<ol className={["a-z-list", className].join(" ")} {...attrs}>
				{children}
			</ol>
		</>
	);
};

AZList.propTypes = {
	alphabet: PropTypes.elementType.isRequired,
	children: PropTypes.node.isRequired,
	className: PropTypes.string
};

export const AZListItem = props => {
	const { id, title, children, className, ...attrs } = props;

	const itemId = id || title.replace(" ", "").toLowerCase();

	return (
		<li className={["a-z-list__item", className].join(" ")} {...attrs}>
			<h2 className="a-z-list__item-heading" id={itemId}>
				{title}
			</h2>
			{children}
		</li>
	);
};

AZListItem.propTypes = {
	id: PropTypes.string,
	title: PropTypes.string.isRequired,
	children: PropTypes.node.isRequired,
	className: PropTypes.string
};
