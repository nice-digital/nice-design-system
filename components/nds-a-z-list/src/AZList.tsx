import React from "react";
import "../scss/a-z-list.scss";

export interface AZListProps {
	alphabet: React.ElementType;
	children: React.ReactNode;
	className?: string;
	[prop: string]: unknown;
}

export interface AZListItemProps {
	id?: string;
	title: string;
	children: React.ReactNode;
	className?: string;
	[prop: string]: unknown;
}

export const AZList: React.FC<AZListProps> = (props: AZListProps) => {
	const { alphabet: Alphabet, children, className, ...attrs } = props;

	return (
		<>
			<Alphabet className="a-z-list__alphabet" />
			<ol
				className={["a-z-list", className].join(" ")}
				data-component="a-z-list"
				{...attrs}
			>
				{children}
			</ol>
		</>
	);
};

export const AZListItem: React.FC<AZListItemProps> = (
	props: AZListItemProps
) => {
	const { id, title, children, className, ...attrs } = props;

	const itemId = id || title.replace(" ", "").toLowerCase();

	return (
		<li className={["a-z-list__item", className].join(" ")} {...attrs}>
			<h2 className="a-z-list__item-heading" id={itemId}>
				{title}
			</h2>
			{children}
		</li>
	);
};
