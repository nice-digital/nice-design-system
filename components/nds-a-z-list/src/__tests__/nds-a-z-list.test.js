"use strict";

import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

import { AZList } from "../AZList";

describe("A-Z List", () => {
	const AZListItems = () => (
		<>
			<li>Item one</li>
			<li>Item two</li>
		</>
	);
	const mockAlphabet = () => (
		<ol>
			<li>
				<a href="#a">A</a>
			</li>
			<li>
				<a href="#b">B</a>
			</li>
		</ol>
	);

	it("should render without crashing", () => {
		const wrapper = shallow(
			<AZList alphabet={mockAlphabet}>
				<AZListItems />
			</AZList>
		);
		expect(wrapper).toHaveLength(1);
	});

	it("should render an ordered list component", () => {
		const wrapper = shallow(
			<AZList alphabet={mockAlphabet}>
				<AZListItems />
			</AZList>
		);
		expect(wrapper.children("ol").length).toEqual(1);
	});

	it("should render an ordered list with the expected class name", () => {
		const wrapper = shallow(
			<AZList alphabet={mockAlphabet}>
				<AZListItems />
			</AZList>
		);
		expect(wrapper.children("ol").hasClass("a-z-list")).toEqual(true);
	});

	it("should append className prop to rendered class attribute for the list", () => {
		const wrapper = shallow(
			<AZList className="test" alphabet={mockAlphabet}>
				<AZListItems />
			</AZList>
		);
		expect(wrapper.children("ol").hasClass("test")).toEqual(true);
	});

	it("should render additional props as attributes for the list", () => {
		const wrapper = shallow(
			<AZList aria-label="test" alphabet={mockAlphabet}>
				<AZListItems />
			</AZList>
		);
		expect(wrapper.children("ol").prop("aria-label")).toEqual("test");
	});

	it("should match snapshot with default attributes", () => {
		const wrapper = mount(
			<AZList alphabet={mockAlphabet}>
				<AZListItems />
			</AZList>
		);
		expect(toJson(wrapper)).toMatchSnapshot();
		wrapper.unmount();
	});

	it("should match snapshot with custom attributes", () => {
		const wrapper = mount(
			<AZList
				className="test-class"
				aria-label="test-label"
				alphabet={mockAlphabet}
			>
				<AZListItems />
			</AZList>
		);
		expect(toJson(wrapper)).toMatchSnapshot();
		wrapper.unmount();
	});

	it("should render an alphabet component that has been passed in", () => {
		const wrapper = mount(
			<AZList alphabet={mockAlphabet}>
				<AZListItems />
			</AZList>
		);
		expect(wrapper.find(mockAlphabet)).toHaveLength(1);
		wrapper.unmount();
	});

	it("should render an alphabet component with a custom class", () => {
		const customAlphabet = () => <div className="test-class">Alphabet</div>;
		const wrapper = mount(
			<AZList alphabet={customAlphabet}>
				<AZListItems />
			</AZList>
		);
		expect(
			wrapper.find(customAlphabet).childAt(0).hasClass("test-class")
		).toEqual(true);
		wrapper.unmount();
	});
});
