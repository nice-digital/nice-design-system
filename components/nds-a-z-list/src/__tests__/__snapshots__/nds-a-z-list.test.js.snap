// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A-Z List should match snapshot with custom attributes 1`] = `
<AZList
  alphabet={[Function]}
  aria-label="test-label"
  className="test-class"
>
  <mockAlphabet
    className="a-z-list__alphabet"
  >
    <ol>
      <li>
        <a
          href="#a"
        >
          A
        </a>
      </li>
      <li>
        <a
          href="#b"
        >
          B
        </a>
      </li>
    </ol>
  </mockAlphabet>
  <ol
    aria-label="test-label"
    className="a-z-list test-class"
  >
    <AZListItems>
      <li>
        Item one
      </li>
      <li>
        Item two
      </li>
    </AZListItems>
  </ol>
</AZList>
`;

exports[`A-Z List should match snapshot with default attributes 1`] = `
<AZList
  alphabet={[Function]}
>
  <mockAlphabet
    className="a-z-list__alphabet"
  >
    <ol>
      <li>
        <a
          href="#a"
        >
          A
        </a>
      </li>
      <li>
        <a
          href="#b"
        >
          B
        </a>
      </li>
    </ol>
  </mockAlphabet>
  <ol
    className="a-z-list "
  >
    <AZListItems>
      <li>
        Item one
      </li>
      <li>
        Item two
      </li>
    </AZListItems>
  </ol>
</AZList>
`;
